File: Memba.WebControls/JavaScript/MultiUpload2.js
/*
 * Copyright © 2005-2008 Memba SA. All rights reserved.
 * 
 * This file is part of Velodoc XP Edition.
 * 
 * Velodoc XP Edition is open-source software: you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the Free Software Foundation,
 * either version 2 of the License, or (at your option) any later version.
 * 
 * Velodoc XP Edition is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License along with Velodoc XP Edition.
 * If not, see <http://www.gnu.org/licenses/>.
 * 
 * For more information, please contact Memba at <http://www.memba.com>.
 * You can find more information about Velodoc at <http://www.velodoc.com> and <http://www.velodoc.com/help>.
 * You can try Velodoc online at <http://www.velodoc.net>.
 *
*/ 
Type.registerNamespace('Memba.WebControls');

//--------------------------------------------
//MultiUpload2EventArgs
//--------------------------------------------
Memba.WebControls.MultiUpload2EventArgs = function(id, value) {

    Sys.Debug.trace("MultiUpload2EventArgs: constructor");

    var e = Function._validateParams(arguments, [
        {name: "id", type: String},
        {name: "value", type: String, mayBeNull: true, optional: true}
    ]);
    if (e) throw e;

    this._id = id;
    this._value = value;
}

Memba.WebControls.MultiUpload2EventArgs.prototype = {

    get_id: function() {
        return this._id;
    },

    get_value: function() {
        return this._value;
    }
}

Memba.WebControls.MultiUpload2EventArgs.registerClass('Memba.WebControls.MultiUpload2EventArgs', null);

//--------------------------------------------
//MultiUpload2 Control
//--------------------------------------------
Memba.WebControls.MultiUpload2 = function(element)
{
    Sys.Debug.trace("MultiUpload2: constructor");
    
    this._Text = null;
    this._CssClass = null;
    this._HoverCssClass = null;
    
    this._LinkSpan = null;
    this._InputsDiv = null;
    this._CurrentInput = null;
    this._CurrentIdNum = null;
    this._Count = null;
    
    Memba.WebControls.MultiUpload2.initializeBase(this, [element]);
}

Memba.WebControls.MultiUpload2.prototype =
{
    get_text : function()
    {
        return this._Text;
    },

    set_text : function(value)
    {
        if (this._Text != value)
        {
            this._Text = value;
            if (this.get_isInitialized() && (this._LinkSpan))
                this._LinkSpan.innerHTML = value;

            // Raise the propertyChanged event
            this.raisePropertyChanged('text');
        }
    },
    
    get_cssClass : function()
    {
        return this._CssClass;
    },
 
    set_cssClass : function(value)
    {
        if (this._CssClass != value)
        {
            this._CssClass = value;
            if (this.get_isInitialized() && (this._LinkSpan))
                this._LinkSpan.className = value;

            // Raise the propertyChanged event
            this.raisePropertyChanged('cssClass');
        }
    },
    
    get_hoverCssClass : function()
    {
        return this._HoverCssClass;
    },
 
    set_hoverCssClass : function(value)
    {
        if (this._HoverCssClass != value)
        {
            this._HoverCssClass = value;
            //if (this.get_isInitialized() && (this._LinkSpan))
            //    this._LinkSpan.className = value;

            // Raise the propertyChanged event
            this.raisePropertyChanged('hoverCssClass');
        }
    },

    get_count : function()
    {
        return this._Count;
    },

    removeInput : function(id)
    {
        Sys.Debug.trace("MultiUpload2: removing input id " + id); 
        var input = $get(id);
        if((input) && (this._InputsDiv) && (this._CurrentInput)
            && (input.id != this._CurrentInput.id) && (input.parentNode.id == this._InputsDiv.id))  
        {
            this._InputsDiv.removeChild(input);
            this._Count--;
        }
        Sys.Debug.trace("MultiUpload2: " + this._Count + " files in multi-upload control");
    },
    
    clear : function()
    {
        Sys.Debug.trace("MultiUpload2: clear"); 
        while((this._InputsDiv.childNodes) && (this._InputsDiv.childNodes.length > 0))
        {
            this._InputsDiv.removeChild(this._InputsDiv.childNodes[0]);
            this._Count--;
        }
        Sys.Debug.trace("MultiUpload2: " + this._Count + " files in multi-upload control");
    },
    
    // browse event
    add_browse: function(handler) {
        /// <summary>Adds a event handler for the browse event.</summary>
        /// <param name="handler" type="Function">The handler to add to the event.</param>
        this.get_events().addHandler("browse", handler);
    },
    remove_browse: function(handler) {
        /// <summary>Removes a event handler for the browse event.</summary>
        /// <param name="handler" type="Function">The handler to remove from the event.</param>
        this.get_events().removeHandler("browse", handler);
    },
    
    initialize : function()
    {
        Sys.Debug.trace("MultiUpload2: initialization");
        
        Memba.WebControls.MultiUpload2.callBaseMethod(this, 'initialize');
        
        var controlEl = this.get_element();
        
        this._LinkSpan = $get(this.get_id() + "_Link");
        this._LinkSpan.className = this._CssClass;
        
        this._InputsDiv = $get(this.get_id() + "_Inputs");
        this._CurrentInput = this.get_element().childNodes[1];
        this._CurrentIdNum = 0;
        this._Count = 0;
        
        //add mouse over/out handlers
        var onmouseover = Function.createDelegate(this, function() { 
            this._LinkSpan.className = this._HoverCssClass;
        });
        $addHandler(this._CurrentInput, "mouseover", onmouseover);

        var onmouseout = Function.createDelegate(this, function() { 
            this._LinkSpan.className = this._CssClass;
        });
        $addHandler(this._CurrentInput, "mouseout", onmouseout);
                            
        //add click handler        
        var onchange = Function.createDelegate(this, function() {
            
            Sys.Debug.trace("MultiUpload2: Opening file dialog");
            
            var clickedId = this._CurrentInput.id;
            var clickedValue = this._CurrentInput.value;
            Sys.Debug.trace("MultiUpload2: current input id is " + clickedId);
            Sys.Debug.trace("MultiUpload2: value is " + clickedValue);
            
            //Note: Each time a user clicks the control, this delegate is called twice:
            //First time on clicking the control, before the open file dialog is displayed
            //Second time on clicking Open or Cancel in the open file dialog
            //Note that the first time value is empty, like when clicking cancel
            //Thus teh following test 
            if (clickedValue != '')
            {
                //Create new file input element
                this._CurrentIdNum++;
                this._Count++;
                //We do not use cloneNode because Netscape copies the value.
                //Neither newInput.name nor newInput.setAttribute("name", ...) work in IE
                //The name should be added in createElement as in the code provided at
                //http://msdn2.microsoft.com/en-us/library/ms534184.aspx
                //See also: http://www.thunderguy.com/semicolon/2005/05/23/setting-the-name-attribute-in-internet-explorer/
                if (Sys.Browser.agent == Sys.Browser.InternetExplorer)
                    newInput = document.createElement('<input name="X"></input>');
                else
                    newInput = document.createElement("input");
                newInput.id = this.get_id() + "_Input_" + this._CurrentIdNum;
                newInput.name = this.get_id() + "_Input_" + this._CurrentIdNum;
                newInput.type = "file";
                newInput.size = "1";
                Sys.Debug.trace("MultiUpload2: new input name is " + newInput.name);
                newInput.style.cssText = this._CurrentInput.style.cssText;
                Sys.Debug.trace("MultiUpload2: new input style is " + newInput.style.cssText);
                controlEl.insertBefore(newInput, this._CurrentInput);
                               
                //Move the current file input
                var oldInput = this._CurrentInput;
                $clearHandlers(oldInput);
                controlEl.removeChild(oldInput);
                this._InputsDiv.appendChild(oldInput);
                
                //Assign the new input element to the current input
                this._CurrentInput = newInput;           
                
                //add mouse over/out handlers
                var onmouseover = Function.createDelegate(this, function() { 
                    this._LinkSpan.className = this._HoverCssClass;
                });
                $addHandler(this._CurrentInput, "mouseover", onmouseover);

                var onmouseout = Function.createDelegate(this, function() { 
                    this._LinkSpan.className = this._CssClass;
                });
                $addHandler(this._CurrentInput, "mouseout", onmouseout);
                
                //TODO: add onchange handler
                $addHandler(this._CurrentInput, "change", onchange);
                
                //Raise the browse event at the end
                var args = new Memba.WebControls.MultiUpload2EventArgs(oldInput.id, oldInput.value);
                Sys.Debug.trace("MultiUpload2: raising event browse event");
                var eventHandler = this.get_events().getHandler("browse");
                if (eventHandler)
                    eventHandler(this, args);
            }
            
            return true; //So that the open file dialog gets opened
        });
        
        $addHandler(this._CurrentInput, "change", onchange);
    },
    
    dispose : function()
    {
        Sys.Debug.trace("MultiUpload2: disposal");
        
        Memba.WebControls.MultiUpload2.callBaseMethod(this, 'dispose');
    }
}

Memba.WebControls.MultiUpload2.registerClass('Memba.WebControls.MultiUpload2', Sys.UI.Control);

if (typeof (Sys) != 'undefined')
{
    Sys.Application.notifyScriptLoaded();
}

File: Memba.WebControls/MultiUpload2.cs
/*
 * Copyright © 2005-2008 Memba SA. All rights reserved.
 * 
 * This file is part of Velodoc XP Edition.
 * 
 * Velodoc XP Edition is open-source software: you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the Free Software Foundation,
 * either version 2 of the License, or (at your option) any later version.
 * 
 * Velodoc XP Edition is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License along with Velodoc XP Edition.
 * If not, see <http://www.gnu.org/licenses/>.
 * 
 * For more information, please contact Memba at <http://www.memba.com>.
 * You can find more information about Velodoc at <http://www.velodoc.com> and <http://www.velodoc.com/help>.
 * You can try Velodoc online at <http://www.velodoc.net>.
 *
*/
using System;
using System.Collections.Generic;
using System.Text;
using System.Web; //AspNetHostingPermission, AspNetHostingPermissionLevel
using System.Web.UI; //Control, IScriptControl, WebResource, Themeable
using System.Web.UI.Design; //UrlEditor
using System.Web.UI.WebControls;
using System.Drawing.Design; //UITypeEditor
using System.ComponentModel; //Category, Description, DefaultProperty
using System.Security.Permissions; //SecurityAction


//See: http://www.asp.net/AJAX/Documentation/Live/tutorials/IScriptControlTutorial1.aspx

[assembly: WebResource("Memba.WebControls.JavaScript.MultiUpload2.js", "text/javascript")]

namespace Memba.WebControls
{
    [AspNetHostingPermission(SecurityAction.LinkDemand, Level = AspNetHostingPermissionLevel.Minimal)]
    [AspNetHostingPermission(SecurityAction.InheritanceDemand, Level = AspNetHostingPermissionLevel.Minimal)]
    [DefaultProperty("Text")]
    [Themeable(true)]
    //[Designer("...")]
    //[ParseChildren(true)]
    //[PersistChildren(false)]
    public class MultiUpload2 : Control, IScriptControl
    {
        private const string BROWSE = "Browse...";

        #region Private Members
        private string _CssClass;
        private string _HoverCssClass;
        private Unit _Height;
        private Unit _Width;
        #endregion

        #region Constructor
        /// <summary>
        /// Constructor
        /// </summary>
        public MultiUpload2()
            : base()
        {
            ViewState["Text"] = BROWSE;
        }
        #endregion

        #region Property Accessors
        /// <summary>
        /// Text displayed in the MultiUpload2 Span
        /// </summary>
        [Category("Appearance")]
        [Description("Text displayed in the MultiUpload2 SPan")]
        [Themeable(false)]
        [Localizable(true)]
        [DefaultValue(BROWSE)]
        public string Text
        {
            get { return ViewState["Text"] == null ? "" : ViewState["Text"].ToString(); }
            set { ViewState["Text"] = value; }
        }
        /// <summary>
        /// CSS class applied to the MultiUpload2 Span
        /// </summary>
        [Category("Appearance")]
        [Description("CSS class applied to the MultiUpload2 Span")]
        [DefaultValue("")]
        public string CssClass
        {
            get { return _CssClass; }
            set { _CssClass = value; }
        }
        /// <summary>
        /// CSS class applied when hovering the MultiUpload2 Span
        /// </summary>
        [Category("Appearance")]
        [Description("CSS class applied when hovering the MultiUpload2 Span")]
        [DefaultValue("")]
        public string HoverCssClass
        {
            get { return _HoverCssClass; }
            set { _HoverCssClass = value; }
        }
        /// <summary>
        /// Width of the MultiUpload2 Div
        /// </summary>
        [Category("Layout")]
        [Description("Width of the MultiUpload2 Div")]
        [DefaultValue(typeof(Unit), "")]
        public Unit Width
        {
            get { return _Width; }
            set { _Width = value; }
        }
        /// <summary>
        /// Height of the MultiUpload2 Div
        /// </summary>
        [Category("Layout")]
        [Description("Height of the MultiUpload2 Div")]
        [DefaultValue(typeof(Unit), "")]
        public Unit Height
        {
            get { return _Height; }
            set { _Height = value; }
        }
        /// <summary>
        /// 
        /// </summary>
        [Browsable(true)]
        public override bool EnableTheming
        {
            get
            {
                return base.EnableTheming;
            }
            set
            {
                base.EnableTheming = value;
            }
        }
        /// <summary>
        /// 
        /// </summary>
        [Browsable(true)]
        public override string SkinID
        {
            get
            {
                return base.SkinID;
            }
            set
            {
                base.SkinID = value;
            }
        }
        #endregion

        #region Control Events
        /// <summary>
        /// 
        /// </summary>
        /// <param name="e"></param>
        protected override void OnPreRender(EventArgs e)
        {
            if (!this.DesignMode)
            {
                // Test for ScriptManager and register if it exists
                ScriptManager objScriptManager = ScriptManager.GetCurrent(Page);

                if (objScriptManager == null)
                    throw new HttpException("A ScriptManager control must exist on the current page.");

                //Register control with ScriptManager
                objScriptManager.RegisterScriptControl(this);
            }

            base.OnPreRender(e); //was on top of the method

        }
        /// <summary>
        /// 
        /// </summary>
        /// <param name="writer"></param>
        protected override void Render(HtmlTextWriter writer)
        {
            // The following only works in IE, because calling the click() method
            //on an input type=file DOM element is not supported in Firefox.

            //See: http://www.quirksmode.org/dom/inputfile.html
            //And: http://www.michaelmcgrady.com/simple.jsp

            /*
            writer.AddAttribute(HtmlTextWriterAttribute.Id, this.ClientID);
            writer.AddAttribute(HtmlTextWriterAttribute.Href, "#");
            if (!String.IsNullOrEmpty(_CssClass))
                writer.AddAttribute(HtmlTextWriterAttribute.Class, _CssClass);
            if (!_Width.IsEmpty)
                writer.AddStyleAttribute(HtmlTextWriterStyle.Width, _Width.ToString());
            if (!_Height.IsEmpty)
                writer.AddStyleAttribute(HtmlTextWriterStyle.Height, _Height.ToString());
            //writer.AddStyleAttribute(HtmlTextWriterStyle.Cursor, "pointer");
            writer.RenderBeginTag(HtmlTextWriterTag.A);
            writer.Write(this.Text);
            writer.RenderEndTag(); //A

            writer.AddAttribute(HtmlTextWriterAttribute.Id, this.ClientID + "_Inputs");
            writer.AddStyleAttribute(HtmlTextWriterStyle.Display, "none");
            //but Safari is supposed to prefer the following to display:none
            //writer.AddStyleAttribute(HtmlTextWriterStyle.Position, "absolute");
            //writer.AddStyleAttribute(HtmlTextWriterStyle.Left, "-1000px");
            writer.RenderBeginTag(HtmlTextWriterTag.Div);

            writer.AddAttribute(HtmlTextWriterAttribute.Id, this.ClientID + "_Input_0");
            writer.AddAttribute(HtmlTextWriterAttribute.Name, this.ClientID + "_Input_0");
            writer.AddAttribute(HtmlTextWriterAttribute.Type, "file");
            writer.RenderBeginTag(HtmlTextWriterTag.Input);
            writer.RenderEndTag(); //Input

            writer.RenderEndTag(); //Div
            */

            //The general idea is to have a file input control over the label or image
            //with opacity = 0 so that it is not displayed but receives the click event
            //when clicking the label or image.

            //<span id="<id>" style="position: relative;">
            //	<span id="<id>_Link" style='position: relative; left: 0px;'>browse...</span>
            //	<input type="file" id="<id>_Input_<key>" name="X" size="1" style="position:absolute; z-index: 1; left:-42px; font-size:16px; opacity:0; filter:alpha(opacity: 0); -moz-opacity:0; -khtml-opacity: 0; cursor:pointer;" />
            //	<span id="<id>_Mask" style="position: absolute; top:-2px; left:-50px; width: 50px; height:28px; z-index:2 ; background-color: transparent; cursor:default;">&nbsp;</span>
            //</span>
            //<div id="<id>_Inputs" style="display:none;"></div>

            //<id> is the span representing the control 
            //<id>_Link is the span containing the text
            //<id>_Input_<key> is the current file input control
            //<id>_Mask masks the textbox portion of the file input conrol so that it does not get the focus
            //<id>_Inputs is a hidden div containing the file input controls which have been used


            writer.AddAttribute(HtmlTextWriterAttribute.Id, this.ClientID);
            writer.AddStyleAttribute(HtmlTextWriterStyle.Position, "relative");
            if (!_Width.IsEmpty)
                writer.AddStyleAttribute(HtmlTextWriterStyle.Width, _Width.ToString());
            if (!_Height.IsEmpty)
                writer.AddStyleAttribute(HtmlTextWriterStyle.Height, _Height.ToString());
            writer.RenderBeginTag(HtmlTextWriterTag.Span); //BEGIN Span <id>

            writer.AddAttribute(HtmlTextWriterAttribute.Id, this.ClientID + "_Link");
            writer.AddStyleAttribute(HtmlTextWriterStyle.Position, "relative");
            writer.AddStyleAttribute(HtmlTextWriterStyle.Left, "0px"); //Just to make sure the CssClass does not break the alignment with the input and the mask
            if (!String.IsNullOrEmpty(_CssClass))
                writer.AddAttribute(HtmlTextWriterAttribute.Class, _CssClass);
            writer.RenderBeginTag(HtmlTextWriterTag.Span); //BEGIN Span <id>_Link
            writer.Write(this.Text);
            writer.RenderEndTag(); //END Span <id>_Link

            writer.AddAttribute(HtmlTextWriterAttribute.Id, this.ClientID + "_Input_0");
            writer.AddAttribute(HtmlTextWriterAttribute.Name, this.ClientID + "_Input_0");
            writer.AddAttribute(HtmlTextWriterAttribute.Type, "file");
            writer.AddAttribute(HtmlTextWriterAttribute.Size, "1");
            writer.AddStyleAttribute(HtmlTextWriterStyle.Position, "absolute");
            writer.AddStyleAttribute(HtmlTextWriterStyle.ZIndex, "1");
            writer.AddStyleAttribute(HtmlTextWriterStyle.Left, "-42px");
            writer.AddStyleAttribute(HtmlTextWriterStyle.FontSize, "16px"); //A little bit larger to make sure the button covers the link
            writer.AddStyleAttribute(HtmlTextWriterStyle.Cursor, "pointer");
            writer.AddStyleAttribute(HtmlTextWriterStyle.Filter, "alpha(opacity:0)");
            writer.AddStyleAttribute("opacity", "0");
            writer.AddStyleAttribute("-moz-opacity", "0");
            writer.AddStyleAttribute("-khtml-opacity", "0");
            writer.AddStyleAttribute("opacity", "0");
            writer.AddStyleAttribute(HtmlTextWriterStyle.Position, "absolute");
            writer.RenderBeginTag(HtmlTextWriterTag.Input); //BEGIN Input <id>_Input_<key>
            writer.RenderEndTag(); //END Input <id>_Input_<key>

            writer.AddAttribute(HtmlTextWriterAttribute.Id, this.ClientID + "_Mask");
            writer.AddStyleAttribute(HtmlTextWriterStyle.Position, "absolute");
            writer.AddStyleAttribute(HtmlTextWriterStyle.ZIndex, "2");
            writer.AddStyleAttribute(HtmlTextWriterStyle.Top, "-2px");
            writer.AddStyleAttribute(HtmlTextWriterStyle.Left, "-50px");
            writer.AddStyleAttribute(HtmlTextWriterStyle.Width, "50px");
            writer.AddStyleAttribute(HtmlTextWriterStyle.Height, "28px");
            writer.AddStyleAttribute(HtmlTextWriterStyle.BackgroundColor, "transparent");
            writer.AddStyleAttribute(HtmlTextWriterStyle.Cursor, "default");
            writer.RenderBeginTag(HtmlTextWriterTag.Span); //BEGIN Span <id>_Mask
            writer.RenderEndTag(); //END Span <id>_Mask

            writer.RenderEndTag(); //END Span <id>

            writer.AddAttribute(HtmlTextWriterAttribute.Id, this.ClientID + "_Inputs");
            writer.AddStyleAttribute(HtmlTextWriterStyle.Display, "none");
            //but Safari is supposed to prefer the following to display:none
            //writer.AddStyleAttribute(HtmlTextWriterStyle.Position, "absolute");
            //writer.AddStyleAttribute(HtmlTextWriterStyle.Left, "-1000px");
            writer.RenderBeginTag(HtmlTextWriterTag.Div);
            writer.RenderEndTag(); //Div <id>_Inputs

            //Register the Ajax extension properties of the control
            if (!this.DesignMode)
            {
                ScriptManager.GetCurrent(this.Page).RegisterScriptDescriptors(this);
            }

            base.Render(writer); //Was on top of method
        }
        #endregion

        #region IScriptControl interface implementation
        /// <summary>
        /// Protected GetScriptReferences which should be overriden n derived classes
        /// </summary>
        /// <returns></returns>
        protected virtual IEnumerable<ScriptReference> GetScriptReferences()
        {
            yield return new ScriptReference(
                this.Page.ClientScript.GetWebResourceUrl(
                    this.GetType(),
                    "Memba.WebControls.JavaScript.MultiUpload2.js"));
        }
        /// <summary>
        /// Protected GetScriptDescriptors which should be overriden n derived classes
        /// </summary>
        /// <returns></returns>
        protected virtual IEnumerable<ScriptDescriptor> GetScriptDescriptors()
        {
            ScriptControlDescriptor objScriptControlDescriptor = new ScriptControlDescriptor("Memba.WebControls.MultiUpload2", ClientID);
            objScriptControlDescriptor.AddProperty("text", this.Text);
            objScriptControlDescriptor.AddProperty("cssClass", this.CssClass);
            objScriptControlDescriptor.AddProperty("hoverCssClass", this.HoverCssClass);

            yield return objScriptControlDescriptor;
        }
        /// <summary>
        /// Public interface for GetScriptReferences
        /// </summary>
        /// <returns></returns>
        IEnumerable<ScriptReference> IScriptControl.GetScriptReferences()
        {
            return GetScriptReferences();
        }
        /// <summary>
        /// Public interface for GetScriptDescriptors
        /// </summary>
        /// <returns></returns>
        IEnumerable<ScriptDescriptor> IScriptControl.GetScriptDescriptors()
        {
            return GetScriptDescriptors();
        }
        #endregion
    }
}
